src/tensor/cache.rs

```diff
@@ -115,89 +296,80 @@ impl<Ptr> TensorCache<Ptr> {
             let items = cache.get_mut(&key).unwrap();
+            items.ignore_drops += 1;
             // unwrap is safe because reuse is only true if there's at least one item,
             // which is also maintained by the block directly below.
-            let allocation = items.pop().unwrap();
-            // If there are no more cached allocations of this size,
-            // remove the entry from the cache.
-            // This is important for correctness, because the presence
-            // of an entry in the cache indicates that there are valid
-            // allocations to use. (see `let reuse = { ... }` above).
-            if items.is_empty() {
-                cache.remove(&key);
-            }
-            Some(allocation)
+            let allocation = items.allocations.pop().unwrap();
+            allocation.check_key(&key);
+            *write!(self.size) -= allocation.size();
+            Some(allocation.into_storage())
         } else {
             None
         }
     }
 
     /// Inserts an allocation into the cache.
-    pub(crate) fn insert<E>(&self, len: usize, allocation: Ptr) {
+    pub(crate) fn insert<E>(&self, allocation: Ptr::Output<E>)
+    where
+        Ptr::Output<E>: CacheStorage<Output<u8> = Ptr>,
+    {
         if !self.is_enabled() {
-            // This is a panic because it's a bug in the library.
-            panic!("Tried to insert into a disabled cache.");
+            return;
         }
 
+        let allocation = CacheWrapper::from_storage(allocation);
         let layout = Layout::new::<E>();
-        let num_bytes = len * std::mem::size_of::<E>();
+        let num_bytes = allocation.size();
         let key = AllocationKey {
             num_bytes,
             size: layout.size(),
             alignment: layout.align(),
         };
-        #[cfg(not(feature = "no-std"))]
-        let mut cache = self.allocations.write().unwrap();
-        #[cfg(feature = "no-std")]
-        let mut cache = self.allocations.write();
+        allocation.check_key(&key);
+        *write!(self.size) += allocation.size();
+        write!(self.drop_queue).push_back(key);
+        let mut cache = write!(self.allocations);
         if let std::collections::btree_map::Entry::Vacant(e) = cache.entry(key) {
-            #[cfg(not(feature = "no-std"))]
-            {
-                e.insert(std::vec![allocation]);
-            }
-            #[cfg(feature = "no-std")]
-            {
-                let mut allocations = Vec::new();
-                allocations.push(allocation);
-                e.insert(allocations);
-            }
+            e.insert(AllocationGroup {
+                allocations: vec![allocation],
+                ignore_drops: 0,
+            });
         } else {
-            cache.get_mut(&key).unwrap().push(allocation);
+            cache.get_mut(&key).unwrap().allocations.push(allocation);
         }
+        std::mem::drop(cache);
+        self.shrink();
```

Comment:
Think its worth only calling this if necessary?

Reply:
I've made a small modification to `shrink` that minimizes the number of RwLock operations done by shrink when no shrinking needs to occur.